### ISSUE TITLE
Not possible to deserialize CustomResource event into object in kotlin

### ISSUE THREAD

**Issue by manusant** [java-operator-sdk/java-operator-sdk#352]
I'm trying to use the Operator-SDK + Quarkus to build an operator in kotlin but i realized that when receiving events from the kubernetes cluster, the operator fails do deserialize and throws this error:

`11:36:38 ERROR [io.fa.ku.cl.ds.in.WatcherWebSocketListener] (OkHttp https://192.168.64.5:8443/...) Unhandled exception encountered in watcher event handler: java.lang.IllegalStateException: com.fasterxml.jackson.databind.JsonMappingException: object is not an instance of declaring class (through reference chain: io.k8s.operator.tomcat.model.resource.Tomcat["status"]->io.k8s.operator.tomcat.model.status.TomcatStatus["readyReplicas"])
	at io.javaoperatorsdk.operator.processing.CustomResourceCache.clone(CustomResourceCache.java:65)
	at io.javaoperatorsdk.operator.processing.CustomResourceCache.getLatestResource(CustomResourceCache.java:52)
	at io.javaoperatorsdk.operator.processing.DefaultEventHandler.executeBufferedEvents(DefaultEventHandler.java:77)
	at io.javaoperatorsdk.operator.processing.DefaultEventHandler.handleEvent(DefaultEventHandler.java:67)
	at io.javaoperatorsdk.operator.processing.event.internal.CustomResourceEventSource.eventReceived(CustomResourceEventSource.java:105)
	at io.javaoperatorsdk.operator.processing.event.internal.CustomResourceEventSource.eventReceived(CustomResourceEventSource.java:20)
	at io.fabric8.kubernetes.client.utils.WatcherToggle.eventReceived(WatcherToggle.java:49)
	at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.eventReceived(AbstractWatchManager.java:145)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$TypedWatcherWebSocketListener.onMessage(WatchConnectionManager.java:83)
	at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:322)
	at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:273)
	at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:209)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: object is not an instance of declaring class (through reference chain: io.k8s.operator.tomcat.model.resource.Tomcat["status"]->io.k8s.operator.tomcat.model.status.TomcatStatus["readyReplicas"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:390)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:349)
	at com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:316)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:778)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4487)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3742)
	at io.javaoperatorsdk.operator.processing.CustomResourceCache.clone(CustomResourceCache.java:62)
	... 18 more
Caused by: java.lang.IllegalArgumentException: object is not an instance of declaring class
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:689)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
	... 27 more`

**Comment by metacosm** [comment]
We haven't tested the SDK with kotlin. Do you have a project that could be used to reproduce the issue?

**Comment by manusant** [comment]
Yes i've reproduced the Tomcat vs WebApps sample in Kotlin + Quarkus just for learning... i believe that the problem must be related with inheritance and how the Jackson Object mapper handles it. 

**This throws the exception when deserializing:**
```kt
@Group("tomcat.operator.io")
@Version("v1")
class Tomcat : CustomResource<TomcatSpec, TomcatStatus>(), Namespaced
```

**While this works, but i was expecting the properties to be picked automatically by the object mapper:**
```kt
@Group("tomcat.operator.io")
@Version("v1")
class Tomcat(var spec: TomcatSpec = TomcatSpec(), var status: TomcatStatus = TomcatStatus()) : CustomResource<TomcatSpec, TomcatStatus>(), Namespaced
```

I also tried to fix the problem with @JasonSubTypes annotation but i wasn't able to put it working
```kt
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "kind")
@JsonSubTypes(JsonSubTypes.Type(value = TomcatSpec::class, name = "spec"), JsonSubTypes.Type(value = TomcatStatus::class, name = "status"))
```

Als, for kotlin we should use the the kotlin module: https://github.com/FasterXML/jackson-module-kotlin but i wasn't able to use it since the used object mapper is "confined" to the operator-sdk core. I believe that a good ffix would be to receive the object mapper from the outside (constructor), and the application using de SDK should be able provide own object mapper or at least be able to configure it or add extensions. 

The problem is here:
```java
public class CustomResourceCache {

  private ObjectMapper objectMapper = new ObjectMapper();
 ...
}
```

My repo is here https://github.com/manusant/k8s-operator. Also i would like to contribute by bringing this repo to join the other samples you have. 

And last but not least, thank you so much for the library (awesome job)

**Comment by metacosm** [comment]
I think we could probably bypass the issue altogether by removing the `ObjectMapper` from `CustomResourceCache` by adding a copy constructor to `CustomResource` so that we can copy CRs directly instead of going through the quite inefficient process of serializing/deserializing objects… What do you think @csviri?

**Comment by metacosm** [comment]
Hmm, thinking about it some more, it wouldn't really be practical and probably wouldn't be much more efficient than going through serialization while also bringing other issues… Another option would be to make `CustomResource` `Cloneable` with a default implementation that would cover most common cases (i.e. spec and status, no other fields). People with more unique CRs would then need to implement their own `clone` method.

**Comment by csviri** [comment]
@metacosm I'm open to do the cloning in a different way. Clonable makes sense, just would be nice have to enforce it somhow. 
The positive side of the current implementation its very transparent. 

@manusant @metacosm yes we could make even an abstraction over the cache, so define it as an interface and make it replacable with differnet implementations, if that makes sense. For now we can have this one impl where we can parametrise the way of cloning.

**Comment by metacosm** [comment]
Having the `CustomResource` implement `Cloneable` is bringing its own bunch of issues, in particular, this means that the spec and status classes also would have to be `Cloneable` to be able to have a deep copy by default. That has quite an impact on existing code. It also means that you wouldn't be able to do `extends CustomResource<Foo, Void>` anymore to specify that your CR doesn't need a status. 

So maybe the solution is indeed to keep using the serialization way and make the mapper customizable. Do you want to take a stab at it, @manusant?

**Comment by metacosm** [comment]
I believe that this is now fixed with #385 merged. Please give it a try and re-open if this doesn't address your issue.

**Comment by manusant** [comment]
Thank you guys for the effort. From the changes i can see that now it's possible to provide the object mapper from the outside. This will allow us to customize it if needed or use jackson extensions like the kotlin one. 
Once again, great job.
